unstable/lens/isomorphism/base.rkt

```diff
@@ -0,0 +1,67 @@
+#lang racket/base
+
+(provide isomorphism-lens
+         isomorphism-lens?
+         isomorphism-lens-inverse
+         isomorphism-lenses
+         )
+(module+ data
+  (provide string->symbol-lens symbol->string-lens
+           number->string-lens string->number-lens
+           list->vector-lens vector->list-lens
+           list->string-lens string->list-lens
+           hash->list-lens list->hash-lens
+           ))
+
+(require racket/match
+         lens/base/main
+         )
+(module+ test
+  (require rackunit (submod ".." data)))
+
+(struct isomorphism-lens (f inv) #:transparent
+  #:methods gen:lens
+  [(define (lens-view lens tgt)
+     ((isomorphism-lens-f lens) tgt))
+   (define (lens-set lens tgt v)
+     ((isomorphism-lens-inv lens) v))])
+
+(define (isomorphism-lens-inverse lens)
+  (match lens
+    [(isomorphism-lens f inv)
+     (isomorphism-lens inv f)]))
+
+(define (isomorphism-lenses f inv)
+  (values (isomorphism-lens f inv)
+          (isomorphism-lens inv f)))
+
+(module+ data
+  (define-values [string->symbol-lens symbol->string-lens]
+    (isomorphism-lenses string->symbol symbol->string))
+  (define-values [number->string-lens string->number-lens]
+    (isomorphism-lenses number->string string->number))
+  (define-values [list->vector-lens vector->list-lens]
+    (isomorphism-lenses list->vector vector->list))
+  (define-values [list->string-lens string->list-lens]
+    (isomorphism-lenses list->string string->list))
+  (define-values [hash->list-lens list->hash-lens]
+    (isomorphism-lenses hash->list make-immutable-hash))
```

Comment:
Hash->list is tricky because it gives the key-value pairs in an unspecified order. I'm not sure off the top of my head if that causes breaks in any of the lens laws.